go-controller/pkg/ovn/base_network_controller_pods.go

```diff
@@ -985,41 +992,126 @@ func (bnc *BaseNetworkController) shouldReleaseDeletedPod(expectedSwitchName, sw
 		return false, nil
 	}
 
-	if !util.PodCompleted(pod) {
+	if bnc.isReleasedPodBeforeStartup(string(pod.UID), nad) {
+		klog.Infof("Completed pod %s/%s was already released for nad %s before startup",
+			pod.Namespace,
+			pod.Name,
+			nad,
+		)
+		return false, nil
+	}
+
+	// for for secondary network IPs allocated from cluster manager, we will
+	// check if other pods are using the same IPs just in case we are processing
+	// events in different order than the remote zone or cluster manager did
+	// (best effort, there can still be issues with this), otherwise we can
+	// release the IP
+	if bnc.allocatesPodAnnotation() {
```

Comment:
So `canReleasePodIPs` had two purposes:

- For local pods, it was used within `ConditionalIPRelease` to prevent the same bug that is being fixed in this PR (that hopefully covers additional corner cases). Note that this method here is used in for local pods only. So it should be not be possible that two local pods are assigned the same IP by the same zone after this PR fix so we don't need the check.
- For remote pods, it provides coverage in case we are processing events in a different order than the pods zone. But we were not doing this check in cases were the IP was assigned in cluster manager and I think we should.

So I think that what is happening here is that I forgot to remove the `!bnc.allocatesPodAnnotation()` check in `canReleasePodIPs`.

Let me know what you think.

Reply:
>"For local pods, it was used within ConditionalIPRelease to prevent the same bug that is being fixed in this PR (that hopefully covers additional corner cases). Note that this method here shouldReleaseDeletedPod is used in for local pods only. So it should be not be possible that two local pods are assigned the same IP by the same zone after this PR fix so we don't need the check."

This PR is about start up and completed pods. The canReleasePodIPs gives us protection during steady state as well. I don't feel comfortable removing that protection since we have had so many bugs on this in the past.

>"For remote pods, it additionally provides coverage in case we are processing events in a different order than the pods zone." 

I think for remote pods, this was just a side effect that I never thought of honestly. 

>"But we were not doing this check in cases were the IP was assigned in cluster manager and I think we should."

Yes we should also run the check in cluster manager as well.

> So I think that what is happening here is that I forgot to remove the !bnc.allocatesPodAnnotation() check in canReleasePodIPs.

I don't think so. Also, this comment on the code is confusing, because you are talking about code executed later in the function (I think) and not if bnc.allocatesPodAnnotation() :

```
// check if other pods are using the same IPs just in case we are processing
	// events in different order than the remote zone or cluster manager did
	// (best effort, there can still be issues with this), otherwise we can
	// release the IP
```

Reply:
> This PR is about start up and completed pods. The canReleasePodIPs gives us protection during steady state as well. I don't feel comfortable removing that protection since we have had so many bugs on this in the past.

I couldn't come up with any other reason why we would end up with duplicate IPs but I will leave the check around then :(

> Yes we should also run the check in cluster manager as well.

I meant doing this check in the zone controllers for allocations done by cluster manager, not doing the check in cluster manager itself. I would like to leave this check out of cluster manager itself, that's new code and is handling this problem differently and also don't see a reason for duplicate IPs there. If we end up having bugs in the future, they might have a different solution.
 
> I don't think so. 

cluster manager could for example de-allocate an IP for a pod A and allocate the same IP to a different pod B. A zone controller could be stuck with some transient error processing pod A and in the meantime process pod B. After that, it could successfully process pod A maybe resulting in removing the IP from places where it should stay for pod B.

For this kind of scenario I think it would be useful to remove the `!bnc.allocatesPodAnnotation()` check from `canReleasePodIPs`.

> Also, this comment on the code is confusing, because you are talking about code executed later in the function (I think) and not if bnc.allocatesPodAnnotation() :
> 
> ```
> // check if other pods are using the same IPs just in case we are processing
> 	// events in different order than the remote zone or cluster manager did
> 	// (best effort, there can still be issues with this), otherwise we can
> 	// release the IP
> ```

I will rephrase.

Reply:
> cluster manager could for example de-allocate an IP for a pod A and allocate the same IP to a different pod B. A zone controller could be stuck with some transient error processing pod A and in the meantime process pod B. After that, it could successfully process pod A maybe resulting in removing the IP from places where it should stay for pod B.

In this example, I think what becomes clear is the original intent of the function was to determine if we could release the ip from the allocator, but now it is being used to release the ip from address sets or whatever for remote pods (not necessarily the same node that is allocating the IP). In that case like you mentioned the order of events is even more unprotected, so we cannot assume we can release the IP. In that case we need to findPodWithIPAddresses and remove the shortcut like you mentioned:

```
	// in certain configurations IP allocation is handled by cluster manager so
	// we can locally release the IPs without checking
	if !bnc.allocatesPodAnnotation() {
		return true, nil
	}
```

are we on the same page?

Reply:
> > cluster manager could for example de-allocate an IP for a pod A and allocate the same IP to a different pod B. A zone controller could be stuck with some transient error processing pod A and in the meantime process pod B. After that, it could successfully process pod A maybe resulting in removing the IP from places where it should stay for pod B.
> 
> In this example, I think what becomes clear is the original intent of the function was to determine if we could release the ip from the allocator, but now it is being used to release the ip from address sets or whatever for remote pods (not necessarily the same node that is allocating the IP). In that case like you mentioned the order of events is even more unprotected, so we cannot assume we can release the IP. In that case we need to findPodWithIPAddresses and remove the shortcut like you mentioned:
> 
> ```
> 	// in certain configurations IP allocation is handled by cluster manager so
> 	// we can locally release the IPs without checking
> 	if !bnc.allocatesPodAnnotation() {
> 		return true, nil
> 	}
> ```
> 
> are we on the same page?

Not sure, bear with me.

> In this example, I think what becomes clear is the original intent of the function was to determine if we could release the ip from the allocator, but now it is being used to release the ip from address sets or whatever for remote pods (not necessarily the same node that is allocating the IP).

Not sure if the original intention, or our understanding of it, matters :P `canReleasePodIPs` was already being used for remote pods (not here, in `removeRemoteZonePodFromNamespaceAddressSet`) before this PR, I guess with second purpose you mentioned above (since remote pods are not allocated) but it's clear I did not consider the `order of events` problem when I added the `!bnc.allocatesPodAnnotation()` check to `canReleasePodIPs` if that's what you meant, I wouldn't have added that shortcut if I had.

> In that case like you mentioned the order of events is even more unprotected, so we cannot assume we can release the IP. 

Not sure I understand why it would be `more unprotected`. The order of events problem is only relevant when the allocation is not done in the zone, so relevant in any zone when the allocation is done by cluster manager, or remote zones otherwise. Not a problem when the allocation is done in the local zone. This method is specific to local pods so the order of events is not relevant here specifically unless the allocation was done in cluster manager.

> In that case we need to findPodWithIPAddresses and remove the shortcut like you mentioned.

Agreed. And I guess that means we are more less in agreement for this comment:

- I will remove the shortcut from `canReleasePodIPs`.
- I will use `canReleasePodIPs` in this method in any case, regardless on where the allocation was done. I will probably use it through the conditional IP release method from the allocator if our purpose here is to keep things as they were for networks allocated in the zone. But for networks allocated in cluster manager I will use `canReleasePodIPs` directly.

Reply:
The reason I say "more unprotected" is because there is added protection when we use canReleasePodIPs in the context of the ConditionalIPRelease in the process that is handling allocation of IPs. This is because even though events may be simultaneous to add multiple pod events, the allocator is locked, so parallel pod events are serialized.

But in the case of a remote node, it is not responsible for allocation. So in this case there is no allocator lock forcing serialization of events, they are all being processed in parallel. So it could be the correct order of operations are:
1. remove addr x from address set (pod 1 completed)
2. add addr x to address set (pod 2 created)

However the remote node, receives the pod 1 and 2 events and instead:
1. add addr x to address set (pod 2 created)
2. remove addr x from address set (pod 1 completed)

canReleasePodIPs will give us some level of protection on delete remote pod, but it isn't bulletproof. This could potentially happen:

1. pod 1 completed event arrives, canReleasePodIPs is executed, verifies it can be removed
2. pod 2 created/updated event arrives, remote pod is added to address set
3. pod 1 thread now calls deletePodFromNamespace

One way to fix this would be to lock adding/deleting to address sets while we execute the search for CanReleasePodIPs, and execute the delete inside of that lock. We used to lock address sets a while ago though and it had terrible performance impact. Maybe we could try a sync map and store in flight addresses to be added there, and in the delete address set operation we check to see if there is a key in that map and if so ignore the delete.

Reply:
> The reason I say "more unprotected" is because there is added protection when we use canReleasePodIPs in the context of the ConditionalIPRelease in the process that is handling allocation of IPs. This is because even though events may be simultaneous to add multiple pod events, the allocator is locked, so parallel pod events are serialized.

OK, although I don't see the specific reason why we would need this now, I might be missing it. I will keep the check in place.

> 
> But in the case of a remote node, it is not responsible for allocation. So in this case there is no allocator lock forcing serialization of events, they are all being processed in parallel. So it could be the correct order of operations are:
> 
> 1. remove addr x from address set (pod 1 completed)
> 2. add addr x to address set (pod 2 created)
> 
> However the remote node, receives the pod 1 and 2 events and instead:
> 
> 1. add addr x to address set (pod 2 created)
> 2. remove addr x from address set (pod 1 completed)
> 
> canReleasePodIPs will give us some level of protection on delete remote pod, but it isn't bulletproof. This could potentially happen:
> 
> 1. pod 1 completed event arrives, canReleasePodIPs is executed, verifies it can be removed
> 2. pod 2 created/updated event arrives, remote pod is added to address set
> 3. pod 1 thread now calls deletePodFromNamespace
> 
> One way to fix this would be to lock adding/deleting to address sets while we execute the search for CanReleasePodIPs, and execute the delete inside of that lock. We used to lock address sets a while ago though and it had terrible performance impact. Maybe we could try a sync map and store in flight addresses to be added there, and in the delete address set operation we check to see if there is a key in that map and if so ignore the delete.

Agreed. As for this PR, the intention was to keep the level of protection that we had, but additionally have it for allocations that happened in cluster manager.

I think we are in agreement so I will proceed with the changes.

Reply:
> canReleasePodIPs will give us some level of protection on delete remote pod, but it isn't bulletproof. This could potentially happen:
> 
> 1. pod 1 completed event arrives, canReleasePodIPs is executed, verifies it can be removed
> 2. pod 2 created/updated event arrives, remote pod is added to address set
> 3. pod 1 thread now calls deletePodFromNamespace
> 
> One way to fix this would be to lock adding/deleting to address sets while we execute the search for CanReleasePodIPs, and execute the delete inside of that lock. We used to lock address sets a while ago though and it had terrible performance impact. Maybe we could try a sync map and store in flight addresses to be added there, and in the delete address set operation we check to see if there is a key in that map and if so ignore the delete.

Oh, and I forgot. One other reason we might not be trying to mitigate this further is that in the scenario above at step 2 we already have a problem. That is, a zone allocating an IP to a pod that is still being considered for a different pod in a remote zone. Which requires a different approach altogether to fix. And if fixed, we would not into this scenario.